code/game/objects/items/storage/storage.dm

```diff
@@ -519,11 +524,37 @@
 /obj/item/storage/attackby(obj/item/I, mob/user, params)
 	. = ..()
 
+	if(length(refill_types)) //Can anything be emptied into this object
+		for(var/typepath in refill_types)
+			if(istype(I, typepath)) //Is I the right type to refill this object
+				return do_refill(I, user)
+
 	if(!can_be_inserted(I))
 		return
-
 	return handle_item_insertion(I, FALSE, user)
 
+///Refills the storage from the refill_types item
+/obj/item/storage/proc/do_refill(obj/item/storage/refiller, mob/user)
+	if(!length(refiller.contents))
+		user.balloon_alert(user, "[refiller] is empty.")
+		return
+
+	if(length(contents) >= max_storage_space)
+		user.balloon_alert(user, "[src] is full.")
+		return
+
+	user.balloon_alert(user, "Refilling.")
+
+	if(!do_after(user, 15, TRUE, src, BUSY_ICON_GENERIC))
+		return
+
+	playsound(user.loc, refill_sound, 15, 1, 6)
+	for(var/obj/item/IM in refiller)
+		if(length(contents) >= max_storage_space)
```

Comment:
max storage space is the volume the object can hold, which is not necessarily the number of slots. i.e. a satchel has a max of 15, which translates to 5 normal sized object, or 15 injectors.